File: src/core/API/ExplorationTree.cpp
/*
 Copyright (c) 2016 Martin Šícho

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdexcept>
#include <iostream>

#include <morphing/operators/AddAtom.hpp>
#include <morphing/operators/RemoveAtom.hpp>
#include <morphing/operators/RemoveBond.hpp>
#include <morphing/operators/MutateAtom.hpp>
#include <morphing/operators/AddBond.hpp>
#include <morphing/operators/InterlayAtom.hpp>
#include <morphing/operators/ContractBond.hpp>
#include <morphing/operators/RerouteBond.hpp>

#include "selectors/fingerprint_selectors.h"
#include "selectors/chemoper_selectors.h"
#include "selectors/simcoeff_selectors.h"

#include "data_structs/ExplorationTree.hpp"
#include "ExplorationTreeImpl.h"
#include "core/data_structs/ExplorationDataImpl.hpp"
#include "MolpherMolImpl.hpp"
#include "operations/FindLeavesOper.hpp"
#include "operations/FindLeavesOperImpl.hpp"
#include "core/chem/fingerprintStrategy/MorganFngpr.hpp"
#include "core/misc/inout.h"
#include "core/chem/simCoefStrategy/TanimotoSimCoef.hpp"
#include "operations/SortMorphsOper.hpp"
#include "operations/ExtendTreeOper.hpp"
#include "operations/PruneTreeOper.hpp"

ExplorationTree::ExplorationTree() : pimpl(new ExplorationTree::ExplorationTreeImpl()) {
    // no action
}

std::shared_ptr<ExplorationTree> ExplorationTree::create(const ExplorationData& data) {
    auto new_tree = std::shared_ptr<ExplorationTree>(new ExplorationTree());
    new_tree->update(data);
    return new_tree;
}

std::shared_ptr<ExplorationTree> ExplorationTree::create(const std::string& sourceMolAsSMILES, const std::string& targetMolAsSMILES) {
    ExplorationData data;
    data.setSource(MolpherMol(sourceMolAsSMILES));
    if (!data.getSource()->isValid()) {
		throw std::runtime_error("source molecule is invalid");
    }
    if (targetMolAsSMILES.empty()) {
        throw std::runtime_error("target molecule SMILES is empty");
    }
    data.setTarget(MolpherMol(targetMolAsSMILES));
    if (!data.getTarget()->isValid()) {
		throw std::runtime_error("target molecule is invalid");
    }
    return create(data);
}

std::shared_ptr<ExplorationTree> ExplorationTree::create(std::shared_ptr<MolpherMol> source, std::shared_ptr<MolpherMol> target) {
    ExplorationData data;
    data.setSource(*source);
    if (target) {
        data.setTarget(*target);
    }
    return create(data);
}

std::shared_ptr<ExplorationTree> ExplorationTree::create(const std::string& filename) {
    auto data = ExplorationData::load(filename);
    return create(*data);
}

std::shared_ptr<ExplorationData> ExplorationTree::asData() const {
    return pimpl->asData();
}

void ExplorationTree::update(const ExplorationData& data) {
    pimpl->updateData(data, shared_from_this());
}

bool ExplorationTree::hasMol(const std::string& canonSMILES) {
    return pimpl->hasMol(canonSMILES);
}

bool ExplorationTree::hasMol(std::shared_ptr<MolpherMol> mol) {
    return pimpl->hasMol(mol);
}

std::shared_ptr<MolpherMol> ExplorationTree::fetchMol(const std::string& canonSMILES) {
    return pimpl->fetchMol(canonSMILES);
}

void ExplorationTree::runOperation(TreeOperation& operation) {
    pimpl->runOperation(operation, shared_from_this());
}

MolVector ExplorationTree::fetchLeaves(bool increase_dist_improve_counter) {
    return pimpl->fetchLeaves(shared_from_this(), increase_dist_improve_counter);
}

void ExplorationTree::generateMorphs() {
    pimpl->generateMorphs(shared_from_this());
}

void ExplorationTree::generateMorphs(const std::vector<std::shared_ptr<MorphCollector> >& collectors) {
    pimpl->generateMorphs(shared_from_this(), collectors);
}

std::vector<std::shared_ptr<MolpherMol> > ExplorationTree::getCandidateMorphs() {
    return pimpl->getCandidateMorphs();
}

std::vector<bool> ExplorationTree::getCandidateMorphsMask() {
    return pimpl->getCandidateMorphsMask();
}

void ExplorationTree::sortMorphs() {
    pimpl->sortMorphs(shared_from_this());
}

void ExplorationTree::filterMorphs(bool verbose_output) {
    pimpl->filterMorphs(shared_from_this(), verbose_output);
}

void ExplorationTree::filterMorphs(FilterMorphsOper::MorphFilters filters, bool verbose_output) {
    pimpl->filterMorphs(filters, shared_from_this(), verbose_output);
}

void ExplorationTree::extend() {
    pimpl->extend(shared_from_this());
}

void ExplorationTree::deleteSubtree(const std::string& canonSMILES, bool descendents_only) {
    pimpl->deleteSubtree(canonSMILES, descendents_only);
}

void ExplorationTree::prune() {
    pimpl->prune(shared_from_this());
}

unsigned ExplorationTree::getGenerationCount() {
    return pimpl->getGenerationCount();
}

bool ExplorationTree::isPathFound() {
    return pimpl->isPathFound();
}

void ExplorationTree::traverse(const std::string& rootSMILES, TraverseCallback& callback) {
    pimpl->traverse(shared_from_this(), rootSMILES, callback);
}

void ExplorationTree::traverse(TraverseCallback& callback) {
    pimpl->traverse(shared_from_this(), callback);
}

void ExplorationTree::save(const std::string& filename) {
    pimpl->save(filename);
}

void ExplorationTree::setCandidateMorphsMask(const std::vector<bool>& new_mask) {
    pimpl->setCandidateMorphsMask(new_mask);
}

void ExplorationTree::setThreadCount(int threadCnt) {
    pimpl->setThreadCount(threadCnt);
}

int ExplorationTree::getThreadCount() {
    return pimpl->getThreadCount();
}

const std::vector<std::shared_ptr<MorphingOperator>> &ExplorationTree::getMorphingOperators() {
    return pimpl->getMorphingOperators();
}

void ExplorationTree::setMorphingOperators(const std::vector<std::shared_ptr<MorphingOperator>> &operators) {
    pimpl->setMorphingOperators(operators);
}

void ExplorationTree::addMorphingOperator(std::shared_ptr<MorphingOperator> operator_) {
    pimpl->addMorphingOperator(operator_);
}

std::shared_ptr<ExplorationTree> ExplorationTree::create(std::shared_ptr<MolpherMol> source) {
    return ExplorationTree::create(source, nullptr);
}

// pimpl

ExplorationTree::ExplorationTreeImpl::ExplorationTreeImpl() :
generationCnt(0)
, threadCnt(0)
, fingerprint(FP_MORGAN)
, simCoeff(SC_TANIMOTO)
{
    // no action
}

void ExplorationTree::ExplorationTreeImpl::updateData(const ExplorationData& data, std::shared_ptr<ExplorationTree> tree)
{
    if (!data.isValid()) {
        throw std::runtime_error("Supplied exploration data are invalid. " 
                "Cannot update this instance.");
    }
    
    bool is_new_tree = false;
    if (treeMap.empty()) {
        is_new_tree = true;
        auto map = data.getTreeMap();
        for (auto& mol : (*map)) {
            auto added_mol = std::make_shared<MolpherMol>(*(mol.second)); // TODO: a simple move should be more efficient and safe here (http://stackoverflow.com/questions/11711034/stdshared-ptr-of-this)
            added_mol->setOwner(tree);
            treeMap.insert(
                std::make_pair(
                    added_mol->getSMILES()
                    , added_mol
                    )
            );
        }
    } else {
//        Cerr("This tree has already been initialized. Only the morphing parameters will be changed.");
    }
    
    if (is_new_tree) {
        candidates.clear();
        auto cndts = data.getCandidates();
        for (auto& mol : (*cndts)) {
            candidates.push_back(std::make_shared<MolpherMol>(*mol)); // TODO: a simple move should be more efficient and safe here (http://stackoverflow.com/questions/11711034/stdshared-ptr-of-this)
        }
        candidatesMask = data.getCandidatesMask();
                
        morphDerivations.clear();
        for (auto& mol_data : data.getDerivationMap()) {
            morphDerivations.insert(mol_data);
        }
        
        source.reset(new MolpherMol(*(data.getSource())));
        
        generationCnt = data.getGenerationCount();
    }
        
    threadCnt = data.getThreadCount();
    fingerprint = data.getFingerprint();
    simCoeff = data.getSimilarityCoefficient();

    AtomLibrary library(AtomLibrary::getDefaultLibrary()); // TODO: user should be allowed to specify their own library
    if (data.getTarget()) {
        target.reset(new MolpherMol(*(data.getTarget())));
        library = AtomLibrary(*target);
    } else {
        target = nullptr;
    }

    chemOperSelectors = data.getChemicalOperators();
    for (auto val : data.getChemicalOperators()) {
        ChemOperSelector selector = (ChemOperSelector) val;
        switch (selector) {
            case OP_ADD_ATOM:
                chemOpers.push_back(std::make_shared<AddAtom>(library));
                break;
            case OP_REMOVE_ATOM:
                chemOpers.push_back(std::make_shared<RemoveAtom>());
                break;
            case OP_ADD_BOND:
                chemOpers.push_back(std::make_shared<AddBond>());
                break;
            case OP_REMOVE_BOND:
                chemOpers.push_back(std::make_shared<RemoveBond>());
                break;
            case OP_MUTATE_ATOM:
                chemOpers.push_back(std::make_shared<MutateAtom>(library));
                break;
            case OP_INTERLAY_ATOM:
                chemOpers.push_back(std::make_shared<InterlayAtom>(library));
                break;
            case OP_BOND_REROUTE:
                chemOpers.push_back(std::make_shared<RerouteBond>());
                break;
            case OP_BOND_CONTRACTION:
                chemOpers.push_back(std::make_shared<ContractBond>());
                break;
            default:
                break;
        }
    }

    params.cntCandidatesToKeep = data.getCntCandidatesToKeep();
    params.cntCandidatesToKeepMax = data.getCntCandidatesToKeepMax();
    params.itThreshold = data.getItThreshold();
    params.cntMaxMorphs = data.getCntMaxMorphs();
    params.distToTargetDepthSwitch = data.getDistToTargetDepthSwitch();
    params.cntMorphsInDepth = data.getCntMorphsInDepth();
    params.cntMorphs = data.getCntMorphs();
    params.minAcceptableMolecularWeight = data.getMinAcceptableMolecularWeight();
    params.maxAcceptableMolecularWeight = data.getMaxAcceptableMolecularWeight();
	params.sascoreMax = data.getSAScoreMax();
    
    auto new_data = this->asData();
    if (!new_data->isValid()) {
        new_data->save("error_snapshot.xml");
        throw std::runtime_error("The tree was created with serious "
                "inconsistencies. Check the 'error_snapshot.xml' file for more details...");
                // TODO: write some code to be able to find out what went wrong
    }
}

std::shared_ptr<ExplorationData> ExplorationTree::ExplorationTreeImpl::asData() const {
    auto data = std::make_shared<ExplorationData>();
    
    for (auto mol : candidates) {
        data->addCandidate(*mol);
    }
    
    data->setCandidatesMask(candidatesMask);
    data->setChemicalOperators(chemOperSelectors);
    data->setFingerprint(fingerprint);
    data->setGenerationCount(generationCnt);
    
    for (auto& item : morphDerivations) {
        data->addToDerivationMap(item.first, item.second);
    }
    
    data->setCntCandidatesToKeep(params.cntCandidatesToKeep);
    data->setCntCandidatesToKeepMax(params.cntCandidatesToKeepMax);
    data->setItThreshold(params.itThreshold);
    data->setCntMaxMorphs(params.cntMaxMorphs);
    data->setDistToTargetDepthSwitch(params.distToTargetDepthSwitch);
    data->setCntMorphsInDepth(params.cntMorphsInDepth);
    data->setCntMorphs(params.cntMorphs);
    data->setMinAcceptableMolecularWeight(params.minAcceptableMolecularWeight);
    data->setMaxAcceptableMolecularWeight(params.maxAcceptableMolecularWeight);
	data->setSAScoreMax(params.sascoreMax);
    
    data->setSimilarityCoefficient(simCoeff);
    data->setSource(*source);
    if (target) {
        data->setTarget(*target);
    }
    data->setThreadCount(threadCnt);
    
    for (auto& item : treeMap) {
        data->addToTreeMap(item.first, *(item.second));
    }
    
    return data;
}

std::shared_ptr<MolpherMol> ExplorationTree::ExplorationTreeImpl::fetchMol(const std::string& canonSMILES) {
    if (hasMol(canonSMILES)) {
        TreeMap::accessor ac;
        treeMap.find(ac, canonSMILES);
        return ac->second;
    } else {
        throw std::runtime_error("Molecule (" + canonSMILES + ") is not present in the tree.");
    }
}

bool ExplorationTree::ExplorationTreeImpl::hasMol(const std::string& canonSMILES) {
    return treeMap.count(canonSMILES) > 0;
}

bool ExplorationTree::ExplorationTreeImpl::hasMol(std::shared_ptr<MolpherMol> mol) {
    return hasMol(mol->getSMILES());
}

void ExplorationTree::ExplorationTreeImpl::runOperation(TreeOperation& operation, std::shared_ptr<ExplorationTree> tree) {
    operation.setTree(tree);
    operation();
}

MolVector ExplorationTree::ExplorationTreeImpl::fetchLeaves(std::shared_ptr<ExplorationTree> tree, bool increase_dist_improve_counter) {
    FindLeavesOper find(increase_dist_improve_counter);
    runOperation(find, tree);
    return find.fetchLeaves();
}

void ExplorationTree::ExplorationTreeImpl::fetchLeaves(std::shared_ptr<ExplorationTree> tree, bool increase_dist_improve_counter, ConcurrentMolVector& ret) {
    FindLeavesOper::FindLeavesOperImpl find(tree, increase_dist_improve_counter);
    find.fetchLeaves(ret);
}

void ExplorationTree::ExplorationTreeImpl::generateMorphs(std::shared_ptr<ExplorationTree> tree) {
    GenerateMorphsOper generate;
    runOperation(generate, tree);
}

void ExplorationTree::ExplorationTreeImpl::generateMorphs(std::shared_ptr<ExplorationTree> tree, const std::vector<std::shared_ptr<MorphCollector> >& collectors) {
    GenerateMorphsOper generate(collectors);
    runOperation(generate, tree);
}

MolVector ExplorationTree::ExplorationTreeImpl::getCandidateMorphs() {
    MolVector ret;
    for (auto morph : candidates) {
        ret.push_back(morph);
    }
    return ret;
}

std::vector<bool> ExplorationTree::ExplorationTreeImpl::getCandidateMorphsMask() {
    return candidatesMask;
}

void ExplorationTree::ExplorationTreeImpl::sortMorphs(std::shared_ptr<ExplorationTree> tree) {
    DefaultSortCallback callback;
    SortMorphsOper sort(callback);
    runOperation(sort, tree);
}

void ExplorationTree::ExplorationTreeImpl::filterMorphs(std::shared_ptr<ExplorationTree> tree, bool verbose_output) {
    FilterMorphsOper filter(verbose_output);
    runOperation(filter, tree);
}

void ExplorationTree::ExplorationTreeImpl::filterMorphs(FilterMorphsOper::MorphFilters filters, std::shared_ptr<ExplorationTree> tree, bool verbose_output) {
    FilterMorphsOper filter(filters, verbose_output);
    runOperation(filter, tree);
}

void ExplorationTree::ExplorationTreeImpl::extend(std::shared_ptr<ExplorationTree> tree) {
    ExtendTreeOper extend;
    runOperation(extend, tree);
}

void ExplorationTree::ExplorationTreeImpl::deleteSubtree(const std::string& canonSMILES, bool descendents_only) {
    if (hasMol(canonSMILES)) {
        TreeMap::accessor acRoot; // root of the subtree to remove
        treeMap.find(acRoot, canonSMILES);
        assert(!acRoot.empty());
        
        if (!descendents_only) {
            if (acRoot->second->getParentSMILES().empty()) {
                throw std::runtime_error("Deleting the absolute root of the tree (" + canonSMILES + ") is not permitted.");
            }

            TreeMap::accessor acRootParent;
            treeMap.find(acRootParent, acRoot->second->getParentSMILES());
            assert(!acRootParent.empty());
            acRootParent->second->removeFromDescendants(acRoot->second->getSMILES());
            
            acRootParent.release();
            acRoot.release();
            
            erase(canonSMILES);
        } else {
            std::set<std::string>::const_iterator it;
            auto descendants = acRoot->second->getDescendants();
            for (it = descendants.begin();
                    it != descendants.end(); it++) {
                acRoot->second->removeFromDescendants(*it);
                erase(*it);
            }
            acRoot->second->setItersWithoutDistImprovement(0);
        }
    } else {
        throw std::runtime_error("Molecule (" + canonSMILES + ") is not present in the tree.");
    }
}

void ExplorationTree::ExplorationTreeImpl::erase(const std::string& canonSMILES) {
    std::deque<std::string> toErase;
    toErase.push_back(canonSMILES);

    while (!toErase.empty()) {
        std::string current = toErase.front();
        toErase.pop_front();

        TreeMap::accessor ac;
        treeMap.find(ac, current);
        assert(!ac.empty());

        std::set<std::string>::const_iterator it;
        auto descendants = ac->second->getDescendants();
        for (it = descendants.begin();
                it != descendants.end(); it++) {
            toErase.push_back(*it);
        }

//        pruned.push_back(current);
        auto erased_mol = ac->second;
        treeMap.erase(ac);
        erased_mol->removeFromTree();
    }
}

void ExplorationTree::ExplorationTreeImpl::prune(std::shared_ptr<ExplorationTree> tree) {
    PruneTreeOper prune;
    runOperation(prune, tree);
}

unsigned ExplorationTree::ExplorationTreeImpl::getGenerationCount() {
    return generationCnt;
}

bool ExplorationTree::ExplorationTreeImpl::isPathFound() {
    if (target) {
        return hasMol(target);
    } else {
        return false;
    }
}

void ExplorationTree::ExplorationTreeImpl::traverse(std::shared_ptr<ExplorationTree> tree, const std::string& rootSMILES, TraverseCallback& callback) {
    TraverseOper traverse(tree, callback, rootSMILES);
    traverse();
}

void ExplorationTree::ExplorationTreeImpl::traverse(std::shared_ptr<ExplorationTree> tree, TraverseCallback& callback) {
    TraverseOper traverse(tree, callback);
    traverse();
}

void ExplorationTree::ExplorationTreeImpl::save(const std::string& filename) {
    auto data = asData();
    data->save(filename);
}

void ExplorationTree::ExplorationTreeImpl::setCandidateMorphsMask(const std::vector<bool>& new_mask) {
    candidatesMask = new_mask;
}

void ExplorationTree::ExplorationTreeImpl::setThreadCount(int threadCnt) {
    this->threadCnt = threadCnt;
}

int ExplorationTree::ExplorationTreeImpl::getThreadCount() {
    return threadCnt;
}

const std::vector<std::shared_ptr<MorphingOperator>> &ExplorationTree::ExplorationTreeImpl::getMorphingOperators() {
	return chemOpers;
}

void ExplorationTree::ExplorationTreeImpl::setMorphingOperators(
        const std::vector<std::shared_ptr<MorphingOperator>> &operators) {
    chemOpers = operators;
}

void ExplorationTree::ExplorationTreeImpl::addMorphingOperator(std::shared_ptr<MorphingOperator> operator_) {
    chemOpers.push_back(operator_);
}

File: src/python/molpher/core/ExplorationTree.py
# Copyright (c) 2016 Martin Sicho
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import molpher
import warnings

from molpher.core.ExplorationData import ExplorationData
from molpher.core.MolpherMol import MolpherMol
from molpher.core._utils import shorten_repr
from molpher.core.morphing import MorphCollector
from molpher.core.morphing.operators import MorphingOperator
from molpher.core.operations import TraverseOper
from molpher.core.operations.TraverseOper import Callback


class ExplorationTree(molpher.swig_wrappers.core.ExplorationTree):
    """
    This a specialized version of the `molpher.swig_wrappers.core.ExplorationTree` proxy class.
    It implements some additional functionality for ease of use from Python.

    ..  attention:: This class has no constructor defined. Use the :meth:`create`
            factory method to obtain instances of this class.

    .. seealso:: `molpher.swig_wrappers.core.ExplorationTree`

    """

    def __repr__(self):
        return shorten_repr(ExplorationTree, self)

    def __init__(self):
        super(ExplorationTree, self).__init__()

    @staticmethod
    def _cast_mols(mols):
        ret = [x for x in mols]
        for morph in ret:
            morph.__class__ = MolpherMol
        return tuple(ret)

    @staticmethod
    def create(tree_data=None, source=None, target=None):
        """
        Create an exploration tree.

        :param tree_data: the morphing parameters (optional if ``source`` and ``target`` are specified)
        :type tree_data: `molpher.swig_wrappers.core.ExplorationData` (or its derived class),
            a `dict` of parameters (in the same format as in :class:`~molpher.core.ExplorationData.ExplorationData`'s constructor)
            or a path to a :term:`XML template` or a :term:`tree snapshot`
        :param source: SMILES of the source molecule or a molecule directly (uses a copy not the instance itself)
        :type source: `str` or :class:`~molpher.core.MolpherMol.MolpherMol`
        :param target: SMILES of the target molecule
        :type target: `str` or :class:`~molpher.core.MolpherMol.MolpherMol`

        ..  note:: When ``tree_data`` is specified, ``source`` and ``target`` are always ignored.

        """

        ret = None
        if tree_data and source and target:
            warnings.warn(
                "Both tree_data and source and target specified. Using the values in parameters..."
                , RuntimeWarning
            )
        if tree_data and (isinstance(tree_data, molpher.wrappers.ExplorationData) or type(tree_data) == str):
            ret = super(ExplorationTree, ExplorationTree).create(tree_data)
        elif tree_data:
            _params = ExplorationData(**tree_data)
            ret = super(ExplorationTree, ExplorationTree).create(_params)
        elif source and target:
            if type(source) == str:
                source = MolpherMol(source)
            if type(target) == str:
                target = MolpherMol(target)
            ret = super(ExplorationTree, ExplorationTree).create(source, target)
        elif source:
            if type(source) == str:
                source = MolpherMol(source)
            ret = super(ExplorationTree, ExplorationTree).create(source)
        else:
            raise AttributeError('Invalid set of parameters specified.')

        if not ret:
            raise RuntimeError('No tree initilized.')

        ret.callback_class = Callback
        ret.__class__ = ExplorationTree

        return ret

    @property
    def params(self):
        """
        A dictionary representing the current :term:`exploration parameters`.

        It is possible to assign a new dictionary (or an instance of
        the `molpher.swig_wrappers.core.ExplorationData` class)
        to update the current parameters.

        .. note:: Only parameters defined in the supplied dictionary are changed
            and if an instance of `molpher.swig_wrappers.core.ExplorationData`
            is supplied only the parameters are read from it (the tree structure
            remains the same).

        :return: current parameters
        :rtype: `dict`
        """
        data = ExplorationData(other=self.asData())
        return data.param_dict

    @params.setter
    def params(self, params):
        if isinstance(params, molpher.wrappers.ExplorationData):
            self.update(params)
        else:
            self_data = self.asData()
            self_data.param_dict = params
            self.update(self_data)

    @property
    def morphing_operators(self):
        ret = self.getMorphingOperators()
        if ret:
            for x in ret:
                x.__class__ = MorphingOperator
        return ret

    @morphing_operators.setter
    def morphing_operators(self, opers):
        self.setMorphingOperators(opers)

    @property
    def generation_count(self):
        """
        :return: Number of :term:`morph generations <morph generation>` connected to the tree so far.
        :rtype: `int`
        """

        return self.getGenerationCount()

    @property
    def mol_count(self):
        def count_mols(mol):
            count_mols.count += 1
        count_mols.count = 0
        self.traverse(count_mols)
        return count_mols.count

    @property
    def path_found(self):
        """
        :return: `True` if the :term:`target molecule` is present in the tree, `False` otherwise.
        :rtype: `bool`
        """

        return self.isPathFound()

    @property
    def leaves(self):
        """
        :return: the current leaves of the tree
        :rtype: `tuple` of :class:`~molpher.core.MolpherMol.MolpherMol` instances
        """

        return tuple(self._cast_mols(self.fetchLeaves()))

    @property
    def candidates(self):
        """
        :return: the :term:`candidate morphs` (:term:`morphs <morph>` generated by a single call to `generateMorphs()`.)
        :rtype: `tuple` of :class:`~molpher.core.MolpherMol.MolpherMol` instances
        """
        return self._cast_mols(self.getCandidateMorphs())

    @property
    def candidates_mask(self):
        """
        A `tuple` of `bool` objects that serve as means of filtering the :term:`candidate morphs`.
        Each :term:`morph` in `candidates` has a `bool` variable assigned to it in this `tuple`
        -- only morphs with `True` at the appropriate position are added to the tree when :meth:`extend`
        is called.

        It can be changed by assigning a new `tuple` or a call to `setCandidateMorphsMask()`.

        :return: currently selected :term:`candidate morphs` represented as a `tuple` of `bool` objects
        :rtype: `tuple`
        """

        return self.getCandidateMorphsMask()

    @candidates_mask.setter
    def candidates_mask(self, mask):
        self.setCandidateMorphsMask(mask)

    @property
    def thread_count(self):
        """
        :return: maximum number of threads this instance will use
        :rtype: `int`
        """

        return self.getThreadCount()

    @thread_count.setter
    def thread_count(self, val):
        self.setThreadCount(val)

    @property
    def source(self):
        return self.params["source"]

    @property
    def target(self):
        return self.params["target"]

    def fetchMol(self, canonSMILES):
        """
        Returns a molecule from the tree using a canonical
        SMILES string.

        Raises a `RuntimeError` if the molecule is not found.

        :param canonSMILES: SMILES string of the molecule to fetch
        :type canonSMILES: `str`
        :return: molecule from a tree
        :rtype: :class:`~molpher.core.MolpherMol.MolpherMol`
        """

        ret = super(ExplorationTree, self).fetchMol(canonSMILES)
        ret.__class__ = MolpherMol
        return ret

    def traverse(self, callback, start_mol=None):
        """
        This method can be used to traverse the whole tree structure (or just a subtree)
        starting from the root to leaves. It takes a callback function that accepts a single required argument
        and traverses the whole tree starting from its root (or root of a specified subtree -- see ``start_mol``) and
        calls the supplied callback with with encountered morph as its parameter.

        :param callback: the callback to call
        :type callback: a callable object that takes a single argument
        :param start_mol: the root of a subtree to explore as canonical SMILES
            or :py:class:`~molpher.core.MolpherMol.MolpherMol` instance
        :type start_mol: `str` or :py:class:`~molpher.core.MolpherMol.MolpherMol`
        """

        if start_mol and type(start_mol) == MolpherMol:
            TraverseOper(self, self.callback_class(callback), start_mol)()
        elif start_mol:
            mol = self.fetchMol(start_mol)
            TraverseOper(self, self.callback_class(callback), mol)()
        else:
            cb = self.callback_class(callback)
            TraverseOper(self, cb)()

    def generateMorphs(self, collectors=None):
        if not collectors:
            super(ExplorationTree, self).generateMorphs()
        else:
            _collectors = []
            for collector in collectors:
                _collectors.append(MorphCollector(collector))
            super(ExplorationTree, self).generateMorphs(_collectors)

    def asData(self):
        """
        :return: the tree as an :class:`~molpher.core.ExplorationData.ExplorationData` instance
        :rtype: :class:`~molpher.core.ExplorationData.ExplorationData`
        """
        return ExplorationData(other=super(ExplorationTree, self).asData())

    def fetchPathTo(self, smiles):
        if self.hasMol(smiles):
            target = self.fetchMol(smiles)

            path = [target]
            while True:
                parent_smiles = path[-1].parent_smiles
                if parent_smiles:
                    path.append(self.fetchMol(parent_smiles))
                else:
                    break

            path.reverse()
            return path
        else:
            return None
